pkg/kubepkg/fetcher.go

```diff
@@ -0,0 +1,311 @@
+			if err != nil {
+				if !os.IsExist(err) {
+					return fmt.Errorf("creating directory to download %s: %w", pkg.Name, err)
+				}
+			}
+
+			switch pkg.Name {
+			case kubernetesCNIPackage:
+				cniURL, err := GetCNIDownloadLink(pkg.Version, arch)
+				if err != nil {
+					return fmt.Errorf("parsing cni url: %w", err)
+				}
+
+				dlPath := filepath.Join(dlRootPath, "kubernetes-cni.tar.gz")
+				if err := downloadBinaryFromURL(cniURL, dlPath, true); err != nil {
+					return fmt.Errorf("downloading cni binaries: %w", err)
+				}
+			case criToolsPackage:
+				criURL, err := GetCRIToolsDownloadLink(pkg.Version, arch)
+				if err != nil {
+					return fmt.Errorf("parsing cri-tools url: %w", err)
+				}
+
+				dlPath := filepath.Join(dlRootPath, "cri-tools.tar.gz")
+				if err := downloadBinaryFromURL(criURL, dlPath, true); err != nil {
+					return fmt.Errorf("downloading cri-tools binaries: %w", err)
+				}
+			default:
+				kubeURL, err := url.JoinPath(pkgBuilder.DownloadLinkBase, arch, pkg.Name)
+				if err != nil {
+					return fmt.Errorf("parsing %s url: %w", pkg.Name, err)
+				}
+
+				dlPath := filepath.Join(dlRootPath, pkg.Name)
+				if err := downloadBinaryFromURL(kubeURL, dlPath, false); err != nil {
+					return fmt.Errorf("downloading %s binaries: %w", pkg.Name, err)
+				}
+			}
+
+			logrus.Infof("Successfully downloaded %s/%s.", arch, pkg.Name)
+		}
+
+		logrus.Infof("Archiving binaries for %s...", pkg.Name)
+
+		archiveSrc := filepath.Join(pkgBuilder.OutputDir, pkg.Name)
+		archiveDst := filepath.Join(pkgBuilder.OutputDir, pkg.Name, fmt.Sprintf("%s_%s.orig.tar.gz", pkg.Name, pkg.Version))
+		if err := createTarGzArchive(archiveSrc, archiveDst); err != nil {
+			return fmt.Errorf("creating archive: %w", err)
+		}
+
+		logrus.Infof("Successfully archived binaries for %s!", pkg.Name)
+	}
+
+	logrus.Info("Binaries for all packages downloaded and archived!")
+
+	return nil
+}
+
+func downloadBinaryFromURL(downloadURL, path string, extractTgz bool) error {
+	out, err := os.Create(path)
+	if err != nil {
+		return fmt.Errorf("creating download destination file: %w", err)
+	}
+	defer out.Close()
+
+	//nolint:gosec // TODO(xmudrii): I don't think we have another way, but let's check.
+	resp, err := http.Get(downloadURL)
+	if err != nil {
+		return fmt.Errorf("downloading binary: %w", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("downloading binary: status code %d", resp.StatusCode)
+	}
+	defer resp.Body.Close()
+
+	if !extractTgz {
+		_, err = io.Copy(out, resp.Body)
+		if err != nil {
+			return fmt.Errorf("writing downloaded binary: %w", err)
+		}
+	} else {
+		if err := extractTarGz(resp.Body, filepath.Dir(path)); err != nil {
+			return fmt.Errorf("extracting .tar.gz archive: %w", err)
+		}
+		if err := os.Remove(path); err != nil {
+			return fmt.Errorf("removing extracted archive: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func extractTarGz(gzipStream io.Reader, path string) error {
```

Comment:
How about using https://github.com/kubernetes-sigs/release-utils/blob/3532635ef5307586ddfc413f4e39af0fd37fa7c7/tar/tar.go#L113 ?

Reply:
The problem with this function is that it expects a file, but this function works directly with the response body without writing it to the disk. Is this a blocking comment or can I resolve it in a follow up?

Reply:
We can surely follow-up on that.